### PR TITLE
Infowindow and tooltip API

### PR DESCRIPTION
**Pull request by alonsogarciapablo**
- Introduces `layer.infowindow` and `layer.infowindow.update(...)`.
- Introduces `layer.tooltip` and `layer.tooltip.update(...)`.
- `createVis` now supports `show_empty_infowindow_fields` option. This option defaults to `false` and determines if empty fields should be included in the infowindows. The idea is that this flag should be set to `true` in the context of the editor:
  - [ ] open a PR with this ☝️  change after we merge this PR.

infowindow-loading

@xavijam would you please take a look at this PR?

cc: @javisantana

### CODE REVIEW

**Comment by alonsogarciapablo** [comment]
retest this please

**Comment by alonsogarciapablo** [comment]
retest this please

File: src/geo/map/cartodb-layer.js [hunk 1]
```diff
@@ -10,13 +12,20 @@ var CartoDBLayer = LayerModelBase.extend({
...
     LayerModelBase.prototype.initialize.apply(this, arguments);
     options = options || {};
 
     this._map = options.map;
     if (attrs && attrs.cartocss) {
       this.set('initialStyle', attrs.cartocss);
     }
+
+    this.infowindow = new InfowindowTemplate(attrs.infowindow);
+    this.unset('infowindow');
```
Comment (xavijam) [inline src/geo/map/cartodb-layer.js]: That's not a bad move 👍 .

File: src/geo/map/cartodb-layer.js [hunk 2]
```diff
@@ -10,13 +12,20 @@ var CartoDBLayer = LayerModelBase.extend({
...
+    attrs = attrs || {};
     LayerModelBase.prototype.initialize.apply(this, arguments);
     options = options || {};
 
     this._map = options.map;
     if (attrs && attrs.cartocss) {
       this.set('initialStyle', attrs.cartocss);
     }
+
+    this.infowindow = new InfowindowTemplate(attrs.infowindow);
```
Comment (xavijam) [inline src/geo/map/cartodb-layer.js]: infowindowView or infowindowModel?

File: src/geo/map/cartodb-layer.js [hunk 3]
```diff
@@ -10,13 +12,20 @@ var CartoDBLayer = LayerModelBase.extend({
...
+    this.unset('infowindow');
+    this.tooltip = new TooltipTemplate(attrs.tooltip);
```
Comment (xavijam) [inline src/geo/map/cartodb-layer.js]: same ^^^ (x2)

File: src/geo/map/cartodb-layer.js (hunk 1 above)
Comment (javisantana) [inline src/geo/map/cartodb-layer.js]: be careful because the unset raises a change...

File: src/geo/map/cartodb-layer.js (hunk 2 above)
Comment (javisantana) [inline src/geo/map/cartodb-layer.js]: the thing is, the inner model creation is often done during the parse

File: src/geo/map/tooltip-template.js [hunk 4]
```diff
@@ -0,0 +1,27 @@
+var _ = require('underscore');
+var Backbone = require('backbone');
+
+var TooltipTemplate = Backbone.Model.extend({
```
Comment (xavijam) [inline src/geo/map/tooltip-template.js]: Backbone.Model in purpose or you are skipping our very own cdb.core.Model :joy:

Comment (xavijam) [inline src/geo/map/tooltip-template.js]: Same above.

File: src/geo/ui/infowindow-model.js [hunk 5]
```diff
@@ -11,178 +10,81 @@ var Backbone = require('backbone');
...
-    this.unset(from);
-  },
-
-  _cloneFields: function() {
-    return _(this.get('fields')).map(function(v) {
-      return _.clone(v);
-    });
+    template: '',
+    template_name: 'infowindow_light',
+    template_type: 'mustache',
```
Comment (xavijam) [inline src/geo/ui/infowindow-model.js]: Maybe a crazy idea, but could it be worth to start using underscore?

File: src/geo/map/cartodb-layer.js (hunk 1 above)
Comment (xavijam) [inline src/geo/map/cartodb-layer.js]: delete this.attributes.infowindow;

File: src/geo/ui/tooltip-view.js [hunk 6]
```diff
@@ -109,11 +109,16 @@ var Tooltip = InfoBox.extend({
...
+
+    var fadeIn = function () {
+      self.$el.fadeIn(100);
+    };
+
+    var fadeOut = function () {
+      self.$el.fadeOut(200);
+    };
+
+    this.showhideTimeout = setTimeout(self._showing ? fadeIn : fadeOut, 50);
```
Comment (xavijam) [inline src/geo/ui/tooltip-view.js]: visibilityTimeout?

File: src/vis/infowindow-manager.js [hunk 7]
```diff
@@ -1,14 +1,14 @@
-var _ = require('underscore');
 var Overlay = require('./vis/overlay');
```
Comment (xavijam) [inline src/vis/infowindow-manager.js]: You love managers :)

File: src/vis/infowindow-manager.js [hunk 8]
```diff
@@ -1,14 +1,14 @@
...
 /**
  * Manages the infowindows for a map. It listens to changes on the collection
  * of layers and binds a new infowindow view/model to CartoDB.js whenever the
  * collection of layers changes
  */
-var InfowindowManager = function (vis) {
+var InfowindowManager = function (vis, options) {
+  options = options || {};
   this._vis = vis;
+  this._showEmptyFields = options.showEmptyFields;
```
Comment (xavijam) [inline src/vis/infowindow-manager.js]: Why not keep use it with `this.options.showEmptyFields`?

File: src/vis/overlays.js [hunk 9]
```diff
@@ -192,17 +192,12 @@ Overlay.register('_header', function (data, vis) {
...
-  }
-
   var infowindowModel = new InfowindowModel({
     template: data.template,
-    template_type: data.templateType,
     alternative_names: data.alternative_names,
     fields: data.fields,
     template_name: data.template_name,
-    template_type: data.template_type
+    template_type: data.templateType // TODO: Remove this?
```
Comment (xavijam) [inline src/vis/overlays.js]: Why do you want to remove it?

**Comment by xavijam** [comment]
Super minor things, 👍 . Looking forward to seeing it in action.

File: src/geo/map/tooltip-template.js (hunk 4 above)
Comment (alonsogarciapablo) [inline src/geo/map/tooltip-template.js]: I don't see the benefits of using cdb.core.Model over using a good old Backbone Model here.... that's why I always go for Backbone.Model...

File: src/geo/map/cartodb-layer.js (hunk 1 above)
Comment (alonsogarciapablo) [inline src/geo/map/cartodb-layer.js]: Good point. I'll do it silently just in case :)

File: src/geo/map/cartodb-layer.js (hunk 2 above)
Comment (alonsogarciapablo) [inline src/geo/map/cartodb-layer.js]: This is the "property" that layers expose as the infowindow API, so I think `.infowindow` is fine. Is not view, nor a model... it's a template! :)

File: src/geo/ui/infowindow-model.js (hunk 5 above)
Comment (alonsogarciapablo) [inline src/geo/ui/infowindow-model.js]: I don't have a preference.

**Pull request merged**